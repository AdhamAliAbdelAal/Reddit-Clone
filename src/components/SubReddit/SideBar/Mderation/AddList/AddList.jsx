import {
  //  Box, ClickAwayListener,
  ListItem, ListItemText,
} from '@mui/material';
import { useState, useEffect } from 'react';
import AddBtn from './Add Bottun/AddBtn';
import {
  AboutContent,
  Lists,
} from './style';

const primary = ['Activism', 'Art', 'Addiction Support', 'Anime', 'Beauty and Makeup', 'Business, Economics, and Finance', 'Careers', 'Cars and Motor Vehicles',
  'Celebrity', 'Crafts and DIY', 'Crypto', 'Culture, Race, and Ethnicity', 'Ethics and Philosophy', 'Family and Relationships', 'Fashion', 'Fitness and Nutrition',
  'Food and Drink', 'Funny/Humor', 'Gamming', 'Gender', 'History', 'Hobbies', 'Home and Garden', 'Home and Garden', 'Learning and Education', 'Law', 'Marketplace and Deals',
  'Mature Themes and Adult Content', 'Medical and Mental Health', "Men's Health", 'Meta/Reddit', 'Outdoors and Nature', 'Militery', 'Moves', 'Music', 'Outdoors and Nature', 'Place',
  'Podcasts and Streamers', 'Polices', 'Progeamming', 'Reading, Writing, and Literature'];
/**
 * AddList(community topics)
 * @component
 * @return {React.Component} - community topics
 */
function AddList(props) {
  const {
    topics, client, Name,
  } = props;
  const [show, setShow] = useState(true);
  const [showList, setShowList] = useState(true);
  // eslint-disable-next-line prefer-const
  const [tags, setTags] = useState(topics);
  const [tempString, setTempString] = useState(tags);
  const [count, setCount] = useState(tags?.length);
  useEffect(() => {
    setTags(topics);
    setTempString(topics);
  }, [topics]);
  // save the change in input field to get count if char
  const handleChange = (event) => {
    console.log((event.target > 'Text').length);
    if (event.target.value.length < 26) {
      setCount((event.target > 'Text').length);
    }
  };
  // handel on press enter to add the community topic
  const handleKeyDown = (e) => {
    if (tempString.length < 25 && !tempString.includes(e.target.value)) {
      if (e.key !== 'Enter') return;
      const { value } = e.target;
      if (!value.trim()) return;
      setTempString([...tempString, value]);
    }
    e.target.value = '';
  };
  // to rempve element when press x
  const removeItem = (value) => {
    setTempString(tempString.filter((ele, index) => index !== value));
  };

  // add item when select from list
  const ListSelected = (e) => {
    e.stopPropagation();
    if (!tempString.includes(e.target.textContent)) {
      setTempString([...tempString, e.target.textContent]);
      const btn = document.getElementById('add');
      btn.click();
      const input = document.getElementById('input');
      input.focus();
    }
  };
  const trueShaw = () => {
    setShow(true);
  };
  const falseShaw = () => {
    setShow(false);
  };
  const trueShawList = () => {
    setShowList(true);
  };
  const setTemp = (t) => {
    setTempString(t);
  };
  const setTag = (t) => {
    setTags(t);
  };
  const sendData = () => {
    client.patch(`subreddit/${Name}`, { topics: tags }); // fetch api
  };
    // return to default view of add list
  const handleClickAway1 = () => {
    setTags(tempString);
    setShow(true);
    setCount(tags.length);
    sendData();
  };
  return (
    <AboutContent>
      <AddBtn sendData={sendData} handleClickAway1={handleClickAway1} setTag={setTag} setTemp={setTemp} trueShaw={trueShaw} trueShawList={trueShawList} handleChange={handleChange} handleKeyDown={handleKeyDown} removeItem={removeItem} falseShaw={falseShaw} show={show} tempString={tempString} tags={tags} count={count} />
      {
          !show
          && tempString?.length < 25 && showList
          && (
            <Lists
              subheader={<li />}
            >
              {[0].map((sectionId) => (
                <li key={`section-${sectionId}`}>
                  <ul>
                    {primary.map((item) => (
                      <ListItem key={`item-${sectionId}-${item}`}>
                        <ListItemText
                          primary={`${item}`}
                          sx={{
                            span: {
                              fontWeight: 700,
                              fontSize: 13,
                              cursor: 'pointer',
                            },
                          }}
                          onClick={ListSelected}
                        />
                      </ListItem>
                    ))}
                  </ul>
                </li>
              ))}
            </Lists>
          )
        }
    </AboutContent>
  );
}
export default AddList;
