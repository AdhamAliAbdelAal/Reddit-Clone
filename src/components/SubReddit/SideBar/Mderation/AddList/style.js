import { List } from '@mui/material';
import { styled } from '@mui/material/styles';

export const AboutContent = styled('h2')({
  // padding: 12,
  margin: 0,
  color: '#1c1c1c',
  fill: '#1c1c1c',
  wordWrap: 'break-word',
  marginBlockStart: 0,
  marginBlockEnd: 0,
  // height: 127,
});

export const Lists = styled(List)({
  width: '100%',
  backgroundColor: 'white',
  overflow: 'auto',
  boxShadow: '0 2px 4px 0 rgba(28,28,28,0.2)',
  '& ul': { padding: 0 },
  '& ul>li': {
    paddingTop: 0,
    '&:hover': {
      backgroundColor: '#0079d3',
      color: 'white',
    },
  },
  position: 'absolute',
  zIndex: 1,
  height: 200,
});
