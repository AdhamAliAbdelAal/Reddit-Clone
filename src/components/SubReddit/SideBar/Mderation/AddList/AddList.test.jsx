import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import {
  fireEvent,
  render, screen,
} from '@testing-library/react';
import AddList from './AddList';

it('open input to add disc', async () => {
  // window.history.pushState({}, '', '/Login');
  render(<AddList />);
  const add = screen.getByTestId('add');
  fireEvent.click(add);
  const input = screen.getByTestId('input');
  expect(input).toBeInTheDocument();
});

// test snapshot
test('test snapshot', async () => {
  const tree = renderer.create(<AddList />).toJSON();
  expect(tree).toMatchSnapshot();
});
