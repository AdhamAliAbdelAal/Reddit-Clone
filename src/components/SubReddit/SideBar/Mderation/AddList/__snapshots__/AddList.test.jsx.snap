// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot 1`] = `
<h2
  className="css-fo6dne"
>
  <span
    className="css-9t27zl"
    data-testid="add"
    id="add"
    onClick={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="MuiBox-root css-p58oka"
    />
  </span>
</h2>
`;
